blog/2021-04-05-workflow-engine-principles.md

```diff
@@ -0,0 +1,308 @@
+
+
+
+Every practical workflow engine uses queues to dispatch tasks to workers (working processes that host those tasks).
+
+## Timer Queue (7:10)
+
+Time is important because: 
+
+- every task invocation should have a timeout
+- the workflow itself can time out
+- you need to be able to execute operations like sleeping for 30 days
+
+So you also need an external timer service or timer queue that durably stores and dispatches these timers.
+
+
+![Designing Workflow Engine-v7-images 014](https://user-images.githubusercontent.com/6764957/113586984-c7ea9b00-9660-11eb-87ab-359b0b0b5d48.png)
+
+
+## Consistency and the importance of Transactions (7:35)
+
+We need to store the state of the workflow so that every time we start a workflow, we:
+
+- create state in the database
+- create a timer
+- pull tasks from the task queue for the WorkflowDefinition to pick up
+- when it gives a list of commands we need to update and create tasks and update the state
+
+**These updates across multiple data sources must be transactional**.
+
+
+![Designing Workflow Engine-v7-images 016](https://user-images.githubusercontent.com/6764957/113587041-d8027a80-9660-11eb-9d30-d21aaa50f685.png)
+
+
+If you run atop a workflow engine that doesn't have transactions across all these components, you will run into all sorts of race conditions:
+
+- if you update state, and put tasks in task queue, and the state update goes through, but the task queue update fails, your system ends up in a state where it thinks there is a task outstanding but the task is not.
+- if you put tasks in a task queue first and then update state, then If your update is slow, the task can be delivered and processed by the time that goes through so it will be inconsistent and need to deal with edge cases
+
+If all these updates are transactional, then all these race conditions disappear. The system becomes simpler and application level code doesn't see all these edge cases and it simplifies the programming model.
+
+This doesn't just apply to the workflow engine implementation. Most engineers don't use workflow-like orchestrators or workflow-engines to write their services. They use queues, databases and other data sources to create hodgepodge architectures with no transactions across them:
+
+![Designing Workflow Engine-v7-images 017](https://user-images.githubusercontent.com/6764957/113587101-e8b2f080-9660-11eb-9841-8f0141844945.png)
+
+Our point about lack of transactions leading to race conditions applies to the majority of ad hoc systems built by developers. 
+
+This is very important to understand: **If you are building a system from these components you're guaranteed to have edge cases in race conditions**. So that's why having a robust engine with something like Temporal's underlying architecture will simplify your life tremendously. If you want to remember one slide for this presentation, remember this one.
```

Comment:
```suggestion
**If you only remember one slide from this presentation, remember this one.**

If you run atop a workflow engine that doesn't have transactions across all these components, you will run into all sorts of race conditions:

- if you update state, and put tasks in task queue, and the state update goes through, but the task queue update fails, your system ends up in a state where it thinks there is a task outstanding but the task is not.
- if you put tasks in a task queue first and then update state, then If your update is slow, the task can be delivered and processed by the time that goes through so it will be inconsistent and need to deal with edge cases

If all these updates are transactional, then all these race conditions disappear. The system becomes simpler and application level code doesn't see all these edge cases and it simplifies the programming model.

This doesn't just apply to the workflow engine implementation. Most engineers don't use workflow-like orchestrators or workflow-engines to write their services. They use queues, databases and other data sources to create hodgepodge architectures with no transactions across them:

![Designing Workflow Engine-v7-images 017](https://user-images.githubusercontent.com/6764957/113587101-e8b2f080-9660-11eb-9841-8f0141844945.png)

Our point about lack of transactions leading to race conditions applies to the majority of ad hoc systems built by developers. 

This is very important to understand: **If you are building a system from these components you're guaranteed to have edge cases in race conditions**. So that's why having a robust engine with something like Temporal's underlying architecture will simplify your life tremendously.
```